### ISSUE TITLE
Mod Browser Crashing - Object Reference Error

### ISSUE THREAD

**Issue by zauberparacelsus** [blushiemagic/tModLoader#70]
Within a few seconds of being opened, the mod browser crashes.

The following output is shown in ModLoader/Logs/Runtime Error.txt:

```
Object reference not set to an instance of an object
  at Terraria.ModLoader.UI.UIMessageBox.DrawSelf (Microsoft.Xna.Framework.Graphics.SpriteBatch spriteBatch) [0x00000] in <filename unknown>:0 
  at Terraria.UI.UIElement.Draw (Microsoft.Xna.Framework.Graphics.SpriteBatch spriteBatch) [0x00000] in <filename unknown>:0 
  at Terraria.UI.UIElement.DrawChildren (Microsoft.Xna.Framework.Graphics.SpriteBatch spriteBatch) [0x00000] in <filename unknown>:0 
  at Terraria.UI.UIElement.Draw (Microsoft.Xna.Framework.Graphics.SpriteBatch spriteBatch) [0x00000] in <filename unknown>:0 
  at Terraria.UI.UIElement.DrawChildren (Microsoft.Xna.Framework.Graphics.SpriteBatch spriteBatch) [0x00000] in <filename unknown>:0 
  at Terraria.UI.UIElement.Draw (Microsoft.Xna.Framework.Graphics.SpriteBatch spriteBatch) [0x00000] in <filename unknown>:0 
  at Terraria.UI.UserInterface.Draw (Microsoft.Xna.Framework.Graphics.SpriteBatch spriteBatch, Microsoft.Xna.Framework.GameTime time) [0x00000] in <filename unknown>:0 
  at Terraria.Main.DrawMenu (Microsoft.Xna.Framework.GameTime gameTime) [0x00000] in <filename unknown>:0 
  at Terraria.Main.do_Draw (Microsoft.Xna.Framework.GameTime gameTime) [0x00000] in <filename unknown>:0 
  at Terraria.Main.DoDraw (Microsoft.Xna.Framework.GameTime gameTime) [0x00000] in <filename unknown>:0 
```

**Comment by gamrguy** [comment]
I'm having the same symptoms on Linux, but with a different error:
```Collection was modified; enumeration operation may not execute.
  at System.Collections.Generic.List`1+Enumerator[Terraria.UI.UIElement].MoveNext () [0x00000] in <filename unknown>:0 
  at Terraria.UI.UIElement.RecalculateChildren () [0x00000] in <filename unknown>:0 
  at Terraria.UI.UIElement.Recalculate () [0x00000] in <filename unknown>:0 
  at Terraria.UI.UIElement.RecalculateChildren () [0x00000] in <filename unknown>:0 
  at Terraria.GameContent.UI.Elements.UIList.RecalculateChildren () [0x00000] in <filename unknown>:0 
  at Terraria.UI.UIElement.Recalculate () [0x00000] in <filename unknown>:0 
  at Terraria.GameContent.UI.Elements.UIList.Recalculate () [0x00000] in <filename unknown>:0 
  at Terraria.GameContent.UI.Elements.UIList.DrawSelf (Microsoft.Xna.Framework.Graphics.SpriteBatch spriteBatch) [0x00000] in <filename unknown>:0 
  at Terraria.UI.UIElement.Draw (Microsoft.Xna.Framework.Graphics.SpriteBatch spriteBatch) [0x00000] in <filename unknown>:0 
  at Terraria.UI.UIElement.DrawChildren (Microsoft.Xna.Framework.Graphics.SpriteBatch spriteBatch) [0x00000] in <filename unknown>:0 
  at Terraria.UI.UIElement.Draw (Microsoft.Xna.Framework.Graphics.SpriteBatch spriteBatch) [0x00000] in <filename unknown>:0 
  at Terraria.UI.UIElement.DrawChildren (Microsoft.Xna.Framework.Graphics.SpriteBatch spriteBatch) [0x00000] in <filename unknown>:0 
  at Terraria.UI.UIElement.Draw (Microsoft.Xna.Framework.Graphics.SpriteBatch spriteBatch) [0x00000] in <filename unknown>:0 
  at Terraria.UI.UIElement.DrawChildren (Microsoft.Xna.Framework.Graphics.SpriteBatch spriteBatch) [0x00000] in <filename unknown>:0 
  at Terraria.UI.UIElement.Draw (Microsoft.Xna.Framework.Graphics.SpriteBatch spriteBatch) [0x00000] in <filename unknown>:0 
  at Terraria.ModLoader.UI.UIModBrowser.Draw (Microsoft.Xna.Framework.Graphics.SpriteBatch spriteBatch) [0x00000] in <filename unknown>:0 
  at Terraria.UI.UserInterface.Draw (Microsoft.Xna.Framework.Graphics.SpriteBatch spriteBatch, Microsoft.Xna.Framework.GameTime time) [0x00000] in <filename unknown>:0 
  at Terraria.Main.DrawMenu (Microsoft.Xna.Framework.GameTime gameTime) [0x00000] in <filename unknown>:0 
  at Terraria.Main.do_Draw (Microsoft.Xna.Framework.GameTime gameTime) [0x00000] in <filename unknown>:0 
  at Terraria.Main.DoDraw (Microsoft.Xna.Framework.GameTime gameTime) [0x00000] in <filename unknown>:0 
```

**Comment by zauberparacelsus** [comment]
I had forgotten to note that this was on Linux.  Namely, 64bit Ubuntu 16.10

**Comment by gamrguy** [comment]
Update on this issue: I can now load up the Mod Browser, but the aforementioned crash occurs if I click the Reload List option.

**Comment by Jofairden** [comment]
@zauberparacelsus @gamrguy , still an issue to date?

**Comment by zauberparacelsus** [comment]
Yes.  I just had the crash a moment ago, but the error output was what @gamrguy was having.

Not fun when the game takes 5 minutes to start up...

**Comment by zauberparacelsus** [comment]
Actually, I should clarify: it is now impossible for me to use the mod browser on Linux.  I crash *every* time I use it.

**Comment by Jofairden** [comment]
Since v0.9.2.2 introduces some changes to our UI, let's wait it out and see what happens for you when using it. You'll have to wait until we release (unless someone else can figure out what's going on)

**Comment by zauberparacelsus** [comment]
The mod browser no longer crashes, but instead just always displays "Mod Browser OFFLINE (Unknown)".

**Comment by zauberparacelsus** [comment]
With v0.9.2.3, the "Mod Browser OFFLINE (Unknown)" problem has vanished, but the mod browser has returned to just crashing with the "Collection was modified; enumeration operation may not execute" error message.

**Comment by Jofairden** [comment]
That exception should occur the first time you try to load the browser, is it happening 100% of the time for you?

**Comment by zauberparacelsus** [comment]
Yeah, the "collection was modified" is happening 100% of the time

**Comment by gamrguy** [comment]
Can confirm that this error still occurs. If it doesn't crash the game, the second time runs fine. Sometimes the error won't appear (at least on my end); in this case, the error usually won't appear again until entering a world and going back to the main menu, upon which there seems to be a 100% crash rate.

**Comment by zauberparacelsus** [comment]
Yeah, I'm seeing what @gamrguy is seeing.  First it crashes, then on the next run it works.

**Comment by SerLizar** [comment]
Is it still a problem on linux with v0.10.0.1? It appears to be working correctly on Mac now.

**Comment by gamrguy** [comment]
I haven't seen this issue since 0.10.0.2, and there haven't been any complaints from users on the Discord either. Perhaps this can finally be closed?

**Comment by Jofairden** [comment]
I'm pretty sure this got fixed in e3fe9e3 after my initial attempt in ebb537b
